### ISSUE TITLE
Making contract Sellable

### ISSUE THREAD

**Issue by Lcressot** [OpenZeppelin/openzeppelin-contracts#2889]
**🧐 Motivation**

Hi, I just submitted my first pull request ever here on openzeppelin (PR #2887) and I have been told to discuss my idea here with an issue :)
My idea is to make contracts sellable. For now, Ownable contracts can only be transfered for free to another owner by the owner himself/herself via Ownalble.transferOwnership.

Owning a contract gives a power on this contract and can also be profitable (selling NFTs for instance, or with built-in royalties). Some owners could be intersted in selling it to another party and doing it the safest way possible, which I want to implement. The contract should of course always stay transferable for free as well.

**📝 Details**

My idea for this module are :

- Owner approves buyer: allow the owner to approve a buyer with given price
- Approved buyers buys: allow an approved buyer to buy the contract (the eth. will go to the owner and the ownership will transfer)
- Owner approves proxy: allow the owner to approve a trusted proxy to transfer the ownership of the contract (as it is done for tokens)
- Appproved Proxy trasnfers ownership: allow an approved proxy to transfer the ownership of the contract (proxy can be marketplace sale contracts)

The difficulties that have raised are the following:
- ERC: I would have liked to implement the ERC interface but then my contract could not be inherited by other contracts also inheriting from an ERC interface as I wish it could.
- Ownable inheritance: I would have liked to inherit from the Ownable module but as far as I understand it would then not be possible to transfer the contract ownership with other methods than the Ownable.transferOwnership method.
This latter method is onlyOwner so I can't make a buyerApprove/buy nor proxyApprove/take set of methods to transfer the ownership. Also, the _setOwner is private so cannot be inheritated (as it shouldn't).
What I am proposing is thus to copy/paste the Ownable code inside the Sellable contract, which make Sellable able to use _setOwner, and make child classes able to use Ownable like if it was inherited.

**Comment by Amxx** [comment]
Hello @Lcressot 

AFAIK this approach, of having an ownable-sellable contract is not widely used nor is it clearly documented. As I already said in #2887, I think this should be the topic of a new ERC. I would love to include that in our codebase at some point, but before that, I think this disserves to be discussed in the [ethereum-magicians.org](ethereum-magicians.org) forum and submitted to the [EIP repository](https://github.com/ethereum/EIPs).

For your point about ownable, I believe #2568 addresses it.

**Comment by Lcressot** [comment]
Hey @Amxx 
Let's open this new ERC discussion then! Should I start it on ethereum-magicians.org forum ?
I agree with #2568 if allowing all inheriting class of Ownable to make ownership transfer is okay, which I am not sure right now.

**Comment by Lcressot** [comment]
@Amxx 
Following the discussion at #2887, thanks for the explanations, indeed I wasn't using the correct terminology.

To answer you, I do not intend to build a contract marketplace myself, I just happened to witness people wanting to buy contracts and I had this idea. I am pretty confident that some day soon a lot of contracts in the industry will be sold.

However it will maybe be done in real money with real contracts stipulating who the smart contracts should be transfered to. But there will probably also still be more intermediate size companies or independant willing to have a safe blockchain way to make a sale. And who knows maybe contracts could someday get traded on the stock market ^^

So my goal was really just to contribute but not for a short term usecase of my own. So I will keep in mind your idea of suggesting it to EIP repo and ethereum magicians, maybe we can do that together as I am still very new to these things :)

**Comment by fulldecent** [comment]
I think the main feature that's missing and that is needed is to make transfer safe. 

0xcert's Framework has a solution for this: Ownable should be two-step. First approve, then take. That's what I like in Compound. 

But against still, buying and selling a contract is a VERY involved process. And I do support a marketplace here. The marketplace will be a lot better if it welcomes any contract, not just ones that adopt this proposed interface.

I am biased a few ways here because I'd like to buy contracts (with no less than 1,000 transactions) and I make money auditing contracts.

**Comment by Lcressot** [comment]
@fulldecent 
Indeed, a modification of Ownable with a two-step ownership transfer would probably be the simplest and most effective way to make all contracts potentially sellable on a marketplace !

**Comment by Amxx** [comment]
@fulldecent Maybe you focus too much on the original PR code.

IMO the discussion quickly moved to "only" add something like [this](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2887#issuecomment-930815532).

**Comment by Amxx** [comment]
Sharing the code here for the record:
```
attract contract OwnableApprovable is Ownable {
    address private _approved;
    
    function approveTransferOwnership(address approved) public onlyOwner {
        _approved = approved;
        // Emit an event
    }
    function acceptTransferOwnership(address newOwner) public {
        require(newOwner != address(0));
        require(_msgSender() == _approved);
        _setOwner(newOwner);
        // Emit an event ?
    }
    function _transferOwnership(address newOwner) internal virtual {
        super._transferOwnership(newOwner);
        delete _approved;
    }
}
```

**Comment by fulldecent** [comment]
Instead I would replace and deprecate `Ownable` because the current behavior is dangerous.

```solidity
interface Ownable {
    event OwnershipTransferred(address indexed previousOwner, address indexed newOwner);
    modifier onlyOwner();
    function nominateOwner(address nominee);
    function acceptOwnership();
    function renounceOwnership();
    function owner() view returns (address);
}

**Comment by Amxx** [comment]
The current implementation is standardized (ERC173) and is expected by users. Please understand that for backward compatibility reason we cannot remove it and replace it with something new.

What we can do is extending it (either in the Ownable contract, or as an extension) and maybe at some point in the future deprecate the current version.

Is the interface you shared standardized / published?

**Comment by Lcressot** [comment]
@fulldecent 
I agree with you, it would be the best solution. Also it has always felt weird to me that someone can send contracts or ownership even if the receiver doesn't want them. In NFT world there is a lot of wallet spamming with this!

@Amxx Would it be possible to temporarily add the ```nominateOwner```and ```acceptOwnership``` functions to Ownable while keeping the current ```transferOwnership```untill it gets deprecated ?

**Comment by Amxx** [comment]
We don't do "temporary additions".

The good thing is, you can build and ship an extension to Ownable now that the fundamental functions are available internally. It should be easy, and it will be compatible with our library so our users can choose to depend on both your code and ours.

**And you can do that today without us.**

But if we provide it, we would be committing to maintaining it for the foreseeable future (so that code that uses doesn't break with an update). Also, we would potentially be splitting the community if what we ship is not a standard, and something else emerges. We don't think our role is to initiate the creation of standard *(even though we sometimes do)* but rather to offer quality implementation of standards that the community, as a whole, supports.

I'll discuss that with other members of the team next week, but I believe they will back me on this.

**Comment by fulldecent** [comment]
```
interface Ownable {} // Deprecated

interface OwnershipClaimable { // No ERC needed yet
    event OwnershipTransferred(address indexed previousOwner, address indexed newOwner);
    modifier onlyOwner();
    function nominateOwner(address nominee);
    function claimOwnership();
    function renounceOwnership();
    function owner() view returns (address);
}
```

**Comment by frangio** [comment]
There seem to be two independent things being discussed here: one is the ability to sell a contract, the other is a Claimable variant of Ownable. I don't understand the relation between those two things.

For the ability to sell a contract, which is the focus of this issue, I agree with prior discussion that it should be based around an existing standard for reuse of existing marketplaces, and it seems that ERC721 is fit for this purpose, so it would be a matter of figuring out how to make ownership of a contract into an ERC721 token. But I think this is already reasonably achievable with what we have available:

```solidity
// SPDX-License-Identifier: MIT
pragma solidity 0.8.9;

import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
import "@openzeppelin/contracts/utils/Address.sol";

contract Ownership is ERC721 {
    uint immutable NFT_ID = 1;
    
    constructor(address initialOwner) ERC721("Ownership", "OWN") {
        _mint(initialOwner, NFT_ID);
    }

    function relay(address target, bytes memory data) public {
        require(msg.sender == ownerOf(NFT_ID));
        Address.functionCall(target, data);
    }
}
```

An `Ownable` contract can then set an instance of this contract as the owner.

There is one problem with the above code: a current owner can frontrun a purchase to transfer or renounce ownership of a contract. A possible way to fix this issue is to reject `relay` of `transferOwnership` and `renounceOwnership`, although there may be other functions that should be rejected as well.

**Comment by Mehduzza420** [comment]
A contract should not be sellable. And it should always be protected from anyone doing a pull request or delegate call for withdrawal upon the owners request, and states that the owner shall receive payment. Especially with multiple dependencies. 
With special cases, I propose that it should not be open to public for learning developers and learning reviewers to play with. 
With multiple actors and malicious people on the web we are never sure who we are dealing with.
